rio_tiler/reader.py

```diff
@@ -0,0 +1,552 @@
+        data = vrt.read(
+            out_shape=(len(indexes), height, width),
+            indexes=indexes,
+            window=out_window,
+            resampling=Resampling[resampling_method],
+        )
+        if ColorInterp.alpha in vrt.colorinterp:
+            idx = vrt.colorinterp.index(ColorInterp.alpha) + 1
+            mask = vrt.read(
+                indexes=idx,
+                out_shape=(height, width),
+                window=out_window,
+                resampling=Resampling[resampling_method],
+                out_dtype="uint8",
+            )
+        else:
+            mask = vrt.dataset_mask(
+                out_shape=(height, width),
+                window=out_window,
+                resampling=Resampling[resampling_method],
+            )
+
+        if force_binary_mask:
+            mask = numpy.where(mask != 0, numpy.uint8(255), numpy.uint8(0))
+
+        if unscale:
+            data = data.astype("float32", casting="unsafe")
+            numpy.multiply(data, vrt.scales[0], out=data, casting="unsafe")
+            numpy.add(data, vrt.offsets[0], out=data, casting="unsafe")
+
+    return data, mask
+
+
+def part(
+    src_dst: Union[DatasetReader, DatasetWriter, WarpedVRT],
+    bounds: Tuple[float, float, float, float],
+    height: int,
+    width: int,
+    padding: int = 0,
+    dst_crs: CRS = constants.WEB_MERCATOR_CRS,
+    bounds_crs: Optional[CRS] = None,
+    minimum_overlap: Optional[float] = None,
+    warp_vrt_option: Dict = {},
+    **kwargs: Any,
+) -> Tuple[numpy.ndarray, numpy.ndarray]:
+    """
+    Read part of an image.
+
+    Attributes
+    ----------
+        src_dst : rasterio.io.DatasetReader
+            rasterio.io.DatasetReader object
+        bounds : tuple
+            Output bounds (left, bottom, right, top) in target crs ("dst_crs").
+        height : int
+            Output height of the array.
+        width : int
+            Output width of the array.
+        padding : int, optional
+            Padding to apply to each edge of the tile when retrieving data
+            to assist in reducing resampling artefacts along edges.
+        dst_crs: CRS or str, optional
+            Target coordinate reference system, default is "epsg:3857".
+        bounds_crs: CRS or str, optional
+            Overwrite bounds coordinate reference system, default is equal
+            to the output CRS (dst_crs).
+        minimum_tile_cover: float, optional
+            Minimum % overlap for which to raise an error with dataset not
+            covering enought of the tile.
+        warp_vrt_option: dict, optional
+            These will be passed to the rasterio.warp.WarpedVRT class.
+        kwargs : Any, optional
+            Additional options to forward to reader._read()
+
+    Returns
+    -------
+        data : numpy ndarray
+        mask: numpy array
+
+    """
+    if not bounds_crs:
+        bounds_crs = dst_crs
+
+    bounds = transform_bounds(bounds_crs, dst_crs, *bounds, densify_pts=21)
```

Comment:
following @kylebarron comment. I guess we should also make this optional, and only use transform_bounds if `bounds_crs != dst_crs` ?

Reply:
Actually, to be clearer, I found that when `bounds_crs == dst_crs`, it only took ~1.3ms, it's only when `bounds_crs != dst_crs` that it took 85ms.

In current `_tile_read`, `transform_bounds` is called twice, where the second is reprojecting to mercator coordinates, and you can see it's only the second that's slow
![image](https://user-images.githubusercontent.com/15164633/77365575-fc243f80-6d1b-11ea-8d07-19ff8872025b.png)

Reply:
Excellent, so we can keep it as it is then. 

Can I ask how are you profiling the function?

Reply:
I'm using AWS X-Ray, and using their Python SDK to [define manual timing blocks](https://docs.aws.amazon.com/xray-sdk-for-python/latest/reference/basic.html#manually-create-segment-subsegment).

I'm timing parts of the underlying libraries by shadowing their function definitions, i.e.
```py
# local import
from cogeo_mosaic_tiler.temp import (
    fetch_and_find_assets
)
```

and have a file `temp.py` locally that's essentially the same as the [original file](https://github.com/developmentseed/cogeo-mosaic/blob/master/cogeo_mosaic/utils.py), but with 
```py
xray_recorder.begin_subsegment('name')
...
xray_recorder.end_subsegment()
```
interspersed throughout. E.g.

```py
def _aws_get_data(key, bucket, client: boto3_session.client = None) -> BinaryIO:
    if not client:
        xray_recorder.begin_subsegment('make session')
        session = boto3_session()
        xray_recorder.end_subsegment()

        xray_recorder.begin_subsegment('make client')
        client = session.client("s3")
        xray_recorder.end_subsegment()

    xray_recorder.begin_subsegment('client.get_object')
    response = client.get_object(Bucket=bucket, Key=key)
    xray_recorder.end_subsegment()
    return response["Body"].read()
```

Then make a new lambda package, deploy, pan the web map a little, and go into the AWS console to see the trace output.

I have a couple screenshots in issues [here](https://github.com/developmentseed/cogeo-mosaic-tiler/issues/4#issue-585993510) and [here](https://github.com/developmentseed/cogeo-mosaic-tiler/issues/3#issuecomment-602299957) (the latter didn't profile within `mosaic_tiler`)